music_assistant/server/providers/tidal/__init__.py

```diff
@@ -0,0 +1,382 @@
+            type=ConfigEntryType.SECURE_STRING,
+            label="Refresh token for Tidal",
+            description="You need to link Music Assistant to your Tidal account.",
+            hidden=True,
+            value=values.get(CONF_REFRESH_TOKEN) if values else None,
+        ),
+        ConfigEntry(
+            key=CONF_EXPIRY_TIME,
+            type=ConfigEntryType.STRING,
+            label="Expiry time of auth token for Tidal",
+            hidden=True,
+            value=values.get(CONF_EXPIRY_TIME) if values else None,
+        ),
+        ConfigEntry(
+            key=CONF_USER_ID,
+            type=ConfigEntryType.STRING,
+            label="Your Tidal User ID",
+            description="This is your unique Tidal user ID.",
+            hidden=True,
+            value=values.get(CONF_USER_ID) if values else None,
+        ),
+    )
+
+
+class TidalProvider(MusicProvider):
+    """Implementation of a Tidal MusicProvider."""
+
+    _tidal_session: TidalSession | None = None
+    _tidal_user_id: str | None = None
+
+    async def handle_setup(self) -> None:
+        """Handle async initialization of the provider."""
+        tidal_config = TidalConfig(quality=TidalQuality.lossless, item_limit=10000, alac=False)
+        tidal_session = TidalSession(config=tidal_config)
+        # load tokens from config and add tidal session to provider instance.
+        self._tidal_session = await validate_token_and_refresh(
+            tidal_session, self.mass, self.config
+        )
+        self._tidal_user_id = self.config.get_value(CONF_USER_ID)
+
+    @property
+    def supported_features(self) -> tuple[ProviderFeature, ...]:
+        """Return the features supported by this Provider."""
+        return (
+            ProviderFeature.LIBRARY_ARTISTS,
+            ProviderFeature.LIBRARY_ALBUMS,
+            ProviderFeature.LIBRARY_TRACKS,
+            ProviderFeature.LIBRARY_PLAYLISTS,
+            ProviderFeature.ARTIST_ALBUMS,
+            ProviderFeature.ARTIST_TOPTRACKS,
+            ProviderFeature.SEARCH,
+            ProviderFeature.LIBRARY_ARTISTS_EDIT,
+            ProviderFeature.LIBRARY_ALBUMS_EDIT,
+            ProviderFeature.LIBRARY_TRACKS_EDIT,
+            ProviderFeature.LIBRARY_PLAYLISTS_EDIT,
+            ProviderFeature.PLAYLIST_CREATE,
+            ProviderFeature.SIMILAR_TRACKS,
+            ProviderFeature.BROWSE,
+            ProviderFeature.PLAYLIST_TRACKS_EDIT,
+        )
+
+    async def search(
+        self, search_query: str, media_types=list[MediaType] | None, limit: int = 5
+    ) -> SearchResults:
+        """Perform search on musicprovider.
+
+        :param search_query: Search query.
+        :param media_types: A list of media_types to include. All types if None.
+        :param limit: Number of items to return in the search (per type).
+        """
+        search_query = search_query.replace("'", "")
+        results = await search(self._tidal_session, search_query, media_types, limit)
```

Comment:
replace the "self._tidal_session" with your helper method that checks the token and returns the session

Reply:
Created a decorator function and wrapped all methods